### PR TITLE
Introduce platform implementations for  SHA-256

### PR DESCRIPTION
**Pull request by tobiasheine**
### Description
For one of the next steps, we will need a `hmac sha-256` function to build to request signature. I did some research with the result that there is no pure kotlin version of that available. Furthermore, when asking this on the Kotlin/Native slack people advocated against pure kotlin hashing for performance reasons.

Given this, I decided to first migrate the `SHA-256` encoding, we copied from KEthereum, to use platform API's. As next step we will introduce a `hmac sha 256 function` using the same tools.

### Considerations
Replace `Sha256.kt` with `jvmMain/Sha256Encoder.kt` and `iOSMain/Sha256Encoder.kt`.
On the `JVM` we use `java.security.MessageDigest` and on native `platform.CoreCrypto.CC_SHA256` to do the hashing. We can use native dependencies through the [kotlin/native cinterop](https://github.com/JetBrains/kotlin-native/blob/master/INTEROP.md) feature.

### Tests
Since we're using native dependencies (`iOSMain/Sha256Encoder`), we need to run the tests for `iOSMain/Sha256Encoder.kt` on an iOS simulator. Currently, there is no support for doing this as part of the multiplatform gradle plugin, so I created an xcode project, duplicated the test for the `SHA256Encoder` from `commonTest` and ported it to swift. We will use this xcode project in a future iteration as ios sample to the library.

### CODE REVIEW

File: ios/Sample/sampleTests/sampleTests.swift [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        super.setUp()
+        // Put setup code here. This method is called before the invocation of each test method in the class.
+    }
+    
+    override func tearDown() {
+        // Put teardown code here. This method is called after the invocation of each test method in the class.
+        super.tearDown()
+    }
+    
+    func testSha256Request() {
```
Comment (tobiasheine) [inline ios/Sample/sampleTests/sampleTests.swift:18]: This is the swift port of `commonTest/Sha256Test`

File: kmp-library/build.gradle [hunk 2]
```diff
@@ -50,8 +50,28 @@ kotlin {
             }
         }
     }
+    task packForXCode(type: Sync) {
```
Comment (tobiasheine) [inline kmp-library/build.gradle:53]: We need this task to generate an xcode framework we can import into the sample ios project.

File: kmp-library/src/iosMain/kotlin/com/novoda/aws/v4/signer/hash/Sha256Encoder.kt [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+import platform.CoreCrypto.CC_SHA256_DIGEST_LENGTH
+import platform.Foundation.NSString
+import platform.Foundation.NSUTF8StringEncoding
+import platform.Foundation.dataUsingEncoding
+import platform.Foundation.dataWithLength
+
+actual object Sha256Encoder {
+
+    @ExperimentalUnsignedTypes
+    actual fun encode(value: String): String {
```
Comment (tobiasheine) [inline kmp-library/src/iosMain/kotlin/com/novoda/aws/v4/signer/hash/Sha256Encoder.kt:12]: This is the kotlin/native implementation of SHA-256 using `platform.CoreCrypto.CC_SHA256` through [object pinning](https://kotlinlang.org/api/latest/jvm/stdlib/kotlinx.cinterop/-pinned/index.html).

File: ios/Sample/Sample/AppDelegate.swift [hunk 4]
```diff
@@ -0,0 +1,46 @@
+//
+//  AppDelegate.swift
+//  sample
+//
+//  Created by Tobias Heine on 16.04.19.
+//  Copyright © 2019 Novoda. All rights reserved.
+//
```
Comment (florianmski) [inline ios/Sample/Sample/AppDelegate.swift]: Do we need this?

File: ios/Sample/Sample/ViewController.swift [hunk 5]
```diff
@@ -0,0 +1,27 @@
+//
+//  ViewController.swift
+//  sample
+//
+//  Created by Tobias Heine on 16.04.19.
+//  Copyright © 2019 Novoda. All rights reserved.
+//
```
Comment (florianmski) [inline ios/Sample/Sample/ViewController.swift]: Same here

File: kmp-library/src/iosMain/kotlin/com/novoda/aws/v4/signer/hash/Sha256Encoder.kt [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+        memScoped {
+            val dataIn = (value as NSString).dataUsingEncoding(NSUTF8StringEncoding.toULong())!!
+            val macOut = platform.Foundation.NSMutableData.dataWithLength(CC_SHA256_DIGEST_LENGTH.convert())!!
```
Comment (florianmski) [inline kmp-library/src/iosMain/kotlin/com/novoda/aws/v4/signer/hash/Sha256Encoder.kt]: probably don't need the full path, `NSMutableData` should work?

File: ios/Sample/sampleTests/sampleTests.swift (hunk 1 above)
Comment (florianmski) [reply ios/Sample/sampleTests/sampleTests.swift:18]: Shouldn't we move the `commonTest/Sha256Test` into the `jvmTest` as well?

Comment (tobiasheine) [reply ios/Sample/sampleTests/sampleTests.swift:18]: No, the idea is that you provide tests in the common module and the plugin executes the tests for each platform. Just that the ios support is still missing for this. Once this has been added we instead want to remove this swift test.

Comment (florianmski) [reply ios/Sample/sampleTests/sampleTests.swift:18]: Oh I see, awesome!

**Review by florianmski — APPROVED** [review]

File: kmp-library/src/commonTest/kotlin/com/novoda/aws/v4/signer/hash/Sha256Test.kt [hunk 7]
```diff
@@ -18,7 +18,7 @@ class Sha256Test {
     @ExperimentalUnsignedTypes
     @Test
     fun `get sha-256 from request`() {
-        val contentHash = Sha256.digestAsHexString(request)
+        val contentHash = Sha256Encoder.encode(request)
```
Comment (JozefCeluch) [inline kmp-library/src/commonTest/kotlin/com/novoda/aws/v4/signer/hash/Sha256Test.kt:21]: Should this test file be renamed to `Sha256EncoderTest`?

Comment (tobiasheine) [reply kmp-library/src/commonTest/kotlin/com/novoda/aws/v4/signer/hash/Sha256Test.kt:21]: Done in 38d0d6f

**Review by JozefCeluch — APPROVED** [review]

**Pull request merged**